README.md
# line-height [![Build status](https://travis-ci.org/twolfson/line-height.svg?branch=master)](https://travis-ci.org/twolfson/line-height)

Calculate line-height of an HTML element (IE6 compatible)

This was created for provide a well-tested module for calculating [line-height][] in pixels for [trunkata][], a line-based truncation library for HTML.

[line-height]: https://developer.mozilla.org/en-US/docs/Web/CSS/line-height
[trunkata]: https://github.com/twolfson/trunkata

## Getting Started
`line-height` is available via the following:

- [npm][npm], `npm install line-height`
- [bower][bower], `bower install line-height`
- [component][component], `component install line-height`
- [Download via HTTP][download]

[npm]: http://npmjs.org/
[bower]: http://bower.io/
[component]: http://component.io/
[download]: https://raw.github.com/twolfson/line-height/master/dist/line-height.js

For `npm` and `component`, you can load it in as follows:

```js
var lineHeight = require('line-height');
```

For `bower` and `http`, you can use vanilla JS
```html
<script src="components/line-height.js"></script>
window.lineHeight; // `line-height` is defined on `window` in camelCase
```

or you can use [AMD][amd]

[amd]: http://wiki.commonjs.org/wiki/Modules/AsynchronousDefinition

```js
require(['line-height'], funtion (lineHeight) { /* code */ });
```

or [CommonJS][commonjs] syntax (see `npm`/`component` section).

[commonjs]: http://wiki.commonjs.org/wiki/Modules/1.0

Once you have the module loaded, you can get the `line-height` of any node in the DOM.

```js
// Calculate the `line-height` of the body
lineHeight(document.body); // 19

// Calculate the `line-height` of an h2
var h2 = document.createElement('h2');
document.body.appendChild(h2);
lineHeight(h2); // 29

// Calculate how many lines tall an element is
var div = document.createElement('div');
div.innerHTML = '<p>1</p><p>2</p>';
(lineHeight(div) / div.offsetHeight); // 2, how trunkata performs its calculations
```

## Donations
Support this project and [others by twolfson][projects] via [donations][support-me]

[projects]: http://twolfson.com/projects
[support-me]: http://twolfson.com/support-me

## Documentation
`line-height` provides a single function.

```js
lineHeight(node);
/**
 * Calculate the `line-height` of a given node
 * @param {HTMLElement} node Element to calculate line height of. Must be in the DOM.
 * @returns {Number} `line-height` of the element in pixels
 */
```

## Solved problems
### `line-height: normal`
In a large amount of browsers, the [computed style][computed-style] for an element's `line-height` is `normal` by default.

If it is specified by any other means (e.g. ancestor has a `line-height` or the element has a `line-height` specified), it is either a [CSS length][css-length].

To solve this problem, we create a vanilla element of the same `nodeName` (e.g. `h2` if it is an `h2`), apply the original element's `font-size`, and return the element `offsetHeight`. This is the `height` of `1 line` of the element (i.e. `line-height`).

[computed-style]: https://developer.mozilla.org/en-US/docs/Web/API/window.getComputedStyle
[css-length]: https://developer.mozilla.org/en-US/docs/Web/CSS/length

### Converting `pt`, `pc`, `in`, `cm`, `mm` to `px`
In most browsers, when the `line-height` is specified in `pt`, `pc`, `in`, `cm` or `mm`, the `computedStyle` value is in the same unit.

To solve this problem, we use the [standard ratios of conversion to pixels][css-length] to make a conversion to pixels.

- 3pt to 4px
- 1pc to 16px
- 1in to 96px
- 2.54cm to 96px
- 25.4mm to 96px

### `numeric` font-size in IE6
In IE6, `numeric` `font-size`s (e.g. `font-size: 2.3`) are returned without a unit.

To solve this problem, we treat this number as an `em` since it is relative as well. To do that, we set the element's style to `"numeric value" + "em"`, compute and save the `font-size`, remove the temporary style. This conversion gives us the unit in `pt` which we know how to deal with from before.

## Development
### Testing
Tests can be run once via:

```bash
npm test
# Or with Karma directly via
# npm run test-karma-single
```

Tests can also be run continuously via:

```bash
npm run test-karma-single
```

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint via [grunt](https://github.com/gruntjs/grunt) and test via `npm test`.

## License
Copyright (c) 2013 Todd Wolfson

Licensed under the MIT license.


CHANGELOG.md
# line-height changelog
0.3.0 - Repaired support for textareas. Fixes #4

0.2.1 - Added new linting and simplified tests

0.2.0 - Moved from testem and Testling to Karma and Travis CI

0.1.1 - Corrected dependency on computed-style

0.1.0 - Initial release


test/utils/assert.js
// Create and expose assertion methods (node assertion messages suck in browser)
module.exports = {
  strictEqual: function (a, b, msg) {
    if (a !== b) {
      msg = msg || ('Assertion error: ' + a + ' !== ' + b);
      throw new Error(msg);
    }
  },
  notEqual: function (a, b) {
    if (a == b) {
      msg = msg || ('Assertion error: ' + a + ' == ' + b);
      throw new Error(msg);
    }
  }
};


test/line-height_test.js
// Load in test dependencies
var lineHeight = require('../lib/line-height.js');
var assert = require('proclaim');
var domify = require('domify');
var cssControls = require('css-controls');

// Create common fixture actions
var styleSheet = cssControls.createStyleSheet();
var testUtils = {
  getLineHeight: function (html) {
    before(function () {
      // Save our line height
      this.node = domify(html);
      document.body.appendChild(this.node);
      this.lineHeight = lineHeight(this.node);

      // Sanity check line height is a number
      assert.strictEqual(typeof this.lineHeight, 'number');
      assert.notEqual(isNaN(this.lineHeight), true);
    });
    after(function cleanup () {
      document.body.removeChild(this.node);
      delete this.node;
      delete this.lineHeight;
    });
  },
  styleBody: function (css) {
    before(function styleBodyFn () {
      document.body.style.cssText = css;
    });
    after(function cleanup () {
      document.body.style.cssText = '';
    });
  },
  addGlobalCss: function (selector, rule) {
    var index;
    before(function addGlobalCssFn () {
      index = cssControls.addRule(styleSheet, selector, rule);
    });
    after(function cleanup () {
      cssControls.removeRule(styleSheet, index);
    });
  }
};

// Basic tests
var _defaultLnHeight;
describe('An unstyled div processed by line-height', function () {
  testUtils.getLineHeight('<div>abc</div>');
  // Save the line height for other tests
  before(function saveDefaultLineHeight () {
    _defaultLnHeight = this.lineHeight;
  });

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });
});

describe('A line-height styled div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 50px;">abc</div>');

  it('has the styled line-height\'s height', function () {
    assert.strictEqual(this.lineHeight, 50);
  });
});

// DEV: Tests and disproves that an element has a constant ratio for its font-size
describe('A font-size styled div processed by line-height', function () {
  testUtils.getLineHeight('<div style="font-size: 50px;">abc</div>');

  it('has the styled line-height\'s height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the its font-size', function () {
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, 50);
  });
});

// Intermediate tests
describe('A percentage line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 150%;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the default', function () {
    // DEV: In IE6, 150% !== default * 1.5; 24 !== 28.5 (19 * 1.5)
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, _defaultLnHeight);
  });
});

describe('A relative line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 3em;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the default', function () {
    // DEV: In IE6, 1.3em !== default * 1.3; 22 !== 24.7 (19 * 1.3)
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, _defaultLnHeight);
  });
});

// DEV: This is redundant but the test name is practical
describe('An absolute line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 50px;">abc</div>');

  it('has a line-height equal to 50px', function () {
    assert.strictEqual(this.lineHeight, 50);
  });
});

describe('A numeric line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 2.3;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the default', function () {
    // DEV: In IE6, 2.3 !== default * 2.3; 37 !== 43.6999... (19 * 2.3)
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, _defaultLnHeight);
  });
});

// Verify ancestor -> descendant works on global styling to node level
describe('An inherit line-height div processed by line-height', function () {
  testUtils.styleBody('line-height: 40px;');
  testUtils.getLineHeight('<div style="line-height: inherit;">abc</div>');

  it('has a line-height equal to the inherited amount', function () {
    assert.strictEqual(this.lineHeight, 40);
  });
});

// Verify ancestor -> descendant works on node to node level
describe('A child in a styled div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 50px;"><div id="child">abc</div></div>');

  it('has a line-height equal to the parent\'s line-height', function () {
    var childNode = document.getElementById('child');
    assert.strictEqual(lineHeight(childNode), 50);
  });
});

// Advanced tests
// Verify more global styling inheritance
describe('A globally styled body and an unstyled div processed by line-height', function () {
  testUtils.styleBody('font-size: 40px;');
  testUtils.getLineHeight('<div>abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the body\'s font-size', function () {
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, 40);
  });
});

// Kitchen sink tests
// Testing a pt unit type explicitly
describe('A pt line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 27pt;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  // DEV: This verifies our conversion is correct
  it('has a line-height of 36px', function () {
    assert.strictEqual(this.lineHeight, 36); // 27 * 4/3
  });
});

// Testing a mm unit type explicitly (IE6)
describe('A mm line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 50.8mm;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  // DEV: This verifies our conversion is correct
  it('has a line-height of 192px', function () {
    assert.strictEqual(this.lineHeight, 192); // 50.8 * 96/25.4
  });
});

// Testing a cm unit type explicitly (IE6)
describe('A cm line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 2.54cm;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  // DEV: This verifies our conversion is correct
  it('has a line-height of 96px', function () {
    assert.strictEqual(this.lineHeight, 96); // 2.54 * 96/2.54
  });
});

// Testing a in unit type explicitly (IE6)
describe('A in line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 2in;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  // DEV: This verifies our conversion is correct
  it('has a line-height of 192px', function () {
    assert.strictEqual(this.lineHeight, 192); // 2 * 96
  });
});

// Testing a pc unit type explicitly (IE6)
describe('A pc line-height div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 2pc;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  // DEV: This verifies our conversion is correct
  it('has a line-height of 32px', function () {
    assert.strictEqual(this.lineHeight, 32); // 2pc * 12pt/1pc * 4px/3pt
  });
});

// Mass test all other unit types
// DEV: Units taken from https://developer.mozilla.org/en-US/docs/Web/CSS/length
function testCssLength(cssLength) {
  describe('A ' + cssLength + ' line-height div processed by line-height', function () {
    testUtils.getLineHeight('<div style="line-height: 20' + cssLength + ';">abc</div>');

    it('has a line-height equal to its height', function () {
      var height = this.node.offsetHeight;
      assert.greaterThanOrEqual(this.lineHeight, height - 1);
      assert.lessThanOrEqual(this.lineHeight, height + 1);
    });
  });
}
var cssLengths = ['em', 'ex', 'ch', 'rem', 'vh', 'vw', 'vmin', 'vmax', 'px', 'mm', 'cm', 'in', 'pt', 'pc', 'mozmm'];
for (var i = 0; i < cssLengths; i += 1) {
  testCssLength(cssLengths[i]);
}

// Verify there is no bleeding between
describe('An em line-height with a pt font div processed by line-height', function () {
  testUtils.getLineHeight('<div style="line-height: 2.5em; font-size: 33pt;">abc</div>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height greater than the default', function () {
    var lnHeight = this.lineHeight;
    assert.greaterThan(lnHeight, _defaultLnHeight);
  });
});

// Verify we return a line-height specific for a the tag type (e.g. h2 over div)
describe('A div-specific font-size style and an h2 processed by line-height', function () {
  testUtils.addGlobalCss('div', 'font-size: 60px;');
  testUtils.getLineHeight('<h2>abc</h2>');

  it('has a line-height equal to its height', function () {
    var height = this.node.offsetHeight;
    assert.strictEqual(this.lineHeight, height);
  });

  it('has a line-height under the div font-size', function () {
    var lnHeight = this.lineHeight;
    assert.lessThan(lnHeight, 50);
  });
});


karma.conf.js
// Karma configuration
// Generated on Thu Feb 09 2017 01:48:01 GMT-0800 (PST)
module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha'],


    // list of files / patterns to load in the browser
    files: [
      'test/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/browser/**/*.js': ['browserify']
    },

    // configure browserify
    // https://github.com/nikku/karma-browserify/tree/v5.1.0#browserify-config
    browserify: {
      debug: true,
      transform: ['brfs']
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  });
};


lib/line-height.js
// Load in dependencies
var computedStyle = require('computed-style');

/**
 * Calculate the `line-height` of a given node
 * @param {HTMLElement} node Element to calculate line height of. Must be in the DOM.
 * @returns {Number} `line-height` of the element in pixels
 */
function lineHeight(node) {
  // Grab the line-height via style
  var lnHeightStr = computedStyle(node, 'line-height');
  var lnHeight = parseFloat(lnHeightStr, 10);

  // If the lineHeight did not contain a unit (i.e. it was numeric), convert it to ems (e.g. '2.3' === '2.3em')
  if (lnHeightStr === lnHeight + '') {
    // Save the old lineHeight style and update the em unit to the element
    var _lnHeightStyle = node.style.lineHeight;
    node.style.lineHeight = lnHeightStr + 'em';

    // Calculate the em based height
    lnHeightStr = computedStyle(node, 'line-height');
    lnHeight = parseFloat(lnHeightStr, 10);

    // Revert the lineHeight style
    if (_lnHeightStyle) {
      node.style.lineHeight = _lnHeightStyle;
    } else {
      delete node.style.lineHeight;
    }
  }

  // If the lineHeight is in `pt`, convert it to pixels (4px for 3pt)
  // DEV: `em` units are converted to `pt` in IE6
  // Conversion ratio from https://developer.mozilla.org/en-US/docs/Web/CSS/length
  if (lnHeightStr.indexOf('pt') !== -1) {
    lnHeight *= 4;
    lnHeight /= 3;
  // Otherwise, if the lineHeight is in `mm`, convert it to pixels (96px for 25.4mm)
  } else if (lnHeightStr.indexOf('mm') !== -1) {
    lnHeight *= 96;
    lnHeight /= 25.4;
  // Otherwise, if the lineHeight is in `cm`, convert it to pixels (96px for 2.54cm)
  } else if (lnHeightStr.indexOf('cm') !== -1) {
    lnHeight *= 96;
    lnHeight /= 2.54;
  // Otherwise, if the lineHeight is in `in`, convert it to pixels (96px for 1in)
  } else if (lnHeightStr.indexOf('in') !== -1) {
    lnHeight *= 96;
  // Otherwise, if the lineHeight is in `pc`, convert it to pixels (12pt for 1pc)
  } else if (lnHeightStr.indexOf('pc') !== -1) {
    lnHeight *= 16;
  }

  // Continue our computation
  lnHeight = Math.round(lnHeight);

  // If the line-height is "normal", calculate by font-size
  if (lnHeightStr === 'normal') {
    // Create a temporary node
    var nodeName = node.nodeName;
    var _node = document.createElement(nodeName);
    _node.innerHTML = '&nbsp;';

    // Set the font-size of the element
    var fontSizeStr = computedStyle(node, 'font-size');
    _node.style.fontSize = fontSizeStr;

    // Append it to the body
    var body = document.body;
    body.appendChild(_node);

    // Assume the line height of the element is the height
    var height = _node.offsetHeight;
    lnHeight = height;

    // Remove our child from the DOM
    body.removeChild(_node);
  }

  // Return the calculated height
  return lnHeight;
}

// Export lineHeight
module.exports = lineHeight;
